### ISSUE TITLE
Torch?

### ISSUE THREAD

**Issue by Ph0rk0z** [Chennesxu/triton-turing#24]
Since torch requires triton, will this work with any version if I just compile it and ignore the dependency or will there be symbol errors? Also what happens to SM80 since I have a mixed SM75/SM86 system.

**Comment by Chennesxu** [comment]
There is no binary coupling to break here. `libtorch_cuda.so` has no `DT_NEEDED` entry for Triton and no unresolved Triton symbols. PyTorch imports Triton as a regular Python package, so overriding the pip version constraint will not cause a linker error.

What can break is the Python API, since it changes across Triton releases. This fork reports version 3.7.0, and I tested both a recent and an older PyTorch release:

* **PyTorch 2.10** (pins Triton 3.6): works. On a Turing GPU, `torch.compile` with this fork compiled successfully and matched eager execution for pointwise ops, softmax, matmul, layernorm, and SDPA.
* **PyTorch 2.4** (pins Triton 3.0): fails. Its Inductor code imports `triton_key` with

  `from triton.compiler.compiler import triton_key`

  but `triton_key` has since moved to `triton.runtime.cache`. The import therefore fails before any kernel is compiled.

So a one-minor-version gap is fine in this case, while a six-minor-version gap is not. In practice, use a recent PyTorch release. The pip warning about `triton==3.0.0` only reflects the package dependency constraint; it does not indicate a binary incompatibility. You can use `--no-deps` or simply ignore that warning if you know which Triton version you want to use.

A mixed SM75/SM86 system also does not require any special handling. Triton selects the target from the current CUDA device at runtime: `get_current_target()` reads the compute capability of the active GPU. As a result, each GPU gets a kernel compiled for its own target, and the corresponding binaries can coexist in the cache.

There is no build-time architecture choice that needs to be coordinated across the two GPUs. All changes in this fork are gated on either `capability == 75` or `NvidiaMmaEncodingAttr::isTuring()`, so SM86 follows the unmodified upstream Ampere path.

For example, compiling the same GEMM for SM75, SM80, and SM86 produces `m16n8k8` MMA instructions on Turing and `m16n8k16` on both Ampere targets, matching upstream behavior.

**Comment by Ph0rk0z** [comment]
Did you add the fp8 handling patch from triton-windows? It works on turing as well. This way compiling FP8 models is possible. I think I'm at torch 2.9 still. Maybe this could give a boost to my compiled image paths.

**Comment by Chennesxu** [comment]
> Did you add the fp8 handling patch from triton-windows? It works on turing as well. This way compiling FP8 models is possible. I think I'm at torch 2.9 still. Maybe this could give a boost to my compiled image paths.

The windows branch already had it — it inherits the triton-windows tree, so that patch came along. Linux main didn't until now. Linux main was missing it until recently.
I verified on a TITAN RTX: every finite E4M3 value converts bit-exactly (the two NaN codes decode to ±480, same as triton-windows), upstream conversion tests pass, and torch.compile of an fp8 dequant matmul that used to fail with `type fp8e4nv not supported` now runs. 

One expectation to set: Turing has no FP8 tensor core, so this is a storage feature. Weights are upcast to FP16 in registers, and the MMA is still FP16. With the same kernel, FP8 is ~1.6× faster than FP16 when the GEMM is weight-bound (M=1–16, N=K=4096). However, it is ~13% slower when the GEMM is compute-bound (2048³).

So FP8 checkpoints now fit and compile, but how much faster your image path gets depends on which GEMMs dominate.

**Comment by Ph0rk0z** [comment]
For me it was always a nice uplift. Neither my ampere nor turing hardware has fp8 support and prices prevent me acquiring some any time soon.

I'm going to have to give it a go and compile. Did you block out any of the AMD stuff? In past triton I had to make patches to makefiles or else it wants you to d/l those additional packages. No clue how the experience is on windows. 

I think in 2.10 they also try to steal the fp8 path and push it to fp32. https://github.com/pytorch/pytorch/pull/184008 which is numerically better but slower. Inductor uses triton almost exclusively so its a thing to watch out for and also a nice source of speedup since these patches all make a dent in any comfyui compiling.

**Comment by Chennesxu** [comment]
The patch keys on **"no native fp8" (`cc < 89`)**, not on Turing specifically, so your Ampere card gets the same software path — I compiled the same kernel for `sm75` / `sm80` / `sm86` / `sm89`, and only `sm89` emits the hardware `cvt`.

AMD needs no patching — the ROCm headers are vendored, so that backend downloads nothing.

Set:

* `TRITON_PTXAS_PATH`
* `TRITON_CUOBJDUMP_PATH`
* `TRITON_NVDISASM_PATH`
* `TRITON_CUDACRT_PATH`
* `TRITON_CUDART_PATH`

to your local CUDA installation, and it won't fetch the NVIDIA toolchain either.

LLVM still downloads unless you already have the pinned build to point `LLVM_SYSPATH` at.

One more that may matter more for image models: **bf16 `tl.dot` also fell back to FMA on Turing**.

```bash
TRITON_SM75_BF16_DOT_AS_F16=1
```

runs it on the fp16 Tensor Core — **12–18× on plain GEMM**.

It's off by default, since operands above `65504` become `inf`.

**Comment by Ph0rk0z** [comment]
For LLVM I likely have to compile it since they took away something and break it for conda on older ubuntu. IIRC some pthread error. I will try the variable too, usually all ops get kicked to FP16/FP32 though already in comfy when BF16 isn't supported.
